strawberryfields/program.py

```diff
@@ -0,0 +1,897 @@
+        # index must be found in the dict
+        if ind not in self.reg_refs:
+            raise RegRefError('Subsystem {} does not exist.'.format(ind))
+        return self.reg_refs[ind]
+
+    def _test_regrefs(self, reg):
+        """Make sure reg is a valid selection of subsystems, convert them to RegRefs.
+
+        A register reference is valid if it is properly recorded in self.reg_refs
+        and has not been deleted. The selection is valid if it contains only
+        valid RegRefs and no subsystem is repeated.
+
+        Args:
+            reg (Iterable[int, RegRef]): subsystem references
+        Returns:
+            list[RegRef]: converted subsystem references
+        Raises:
+            .RegRefError: if an invalid subsystem reference is found
+        """
+        temp = []
+        for rr in reg:
+            # must be either an integer or a RegRef
+            if isinstance(rr, RegRef):
+                # regref must be found in the dict values (the RegRefs are compared using __eq__, which, since we do not define it, defaults to "is")
+                if rr not in self.reg_refs.values():
+                    raise RegRefError('Unknown RegRef.')
+                if self.reg_refs[rr.ind] is not rr:
+                    raise RegRefError('Should never happen!')
+            elif isinstance(rr, numbers.Integral):
+                rr = self._index_to_regref(rr)
+            else:
+                raise RegRefError('Subsystems can only be indexed using integers and RegRefs.')
+
+            if not rr.active:
+                raise RegRefError('Subsystem {} has already been deleted.'.format(rr.ind))
+            if rr in temp:
+                raise RegRefError('Trying to act on the same subsystem more than once.')
+            temp.append(rr)
+        return temp
+
+    def append(self, op, reg):
+        """Append a command to the program.
+
+        Args:
+            op (Operation): quantum operation
+            reg (list[int, RegRef]): register subsystem(s) to apply it to
+        Returns:
+            list[RegRef]: subsystem list as RegRefs
+        """
+        if self.locked:
+            raise RuntimeError('The Program is locked, no more Commands can be appended to it.')
+
+        # test that the target subsystem references are ok
+        reg = self._test_regrefs(reg)
+        # also test possible RegRef dependencies
+        self._test_regrefs(op.extra_deps)
+        for rr in reg:
+            # it's used now
+            self.unused_indices.discard(rr.ind)
+        self.circuit.append(Command(op, reg))
+        return reg
+
+    def compile(self, backend='fock', optimize=True):
+        """Compile the program for the given backend.
+
+        The compilation step validates the program, making sure all the Operations used are accepted by the backend.
+        Additionally it may decompose certain gates into sequences of simpler gates.
+
+        TODO for now the compilation happens in-place, i.e. it modifies self.
```

Comment:
Was this because there was an issue with getting `prog_new = prog.compile('chip0')` to work?

Reply:
This turned out to be a bit simpler, and one can always make an explicit copy of the original program if you want to keep it. I'm not opposed to changing it though.

Reply:
I'm on the fence, but lean towards whatever seems clearer from the user perspective. i.e., if you see this,

```python
prog.compile('chip')
```

what would the user expect the behaviour to be - in place, or return a new program?

Put another way, which use case is _less_ surprising to the user? @co9olguy, thoughts?

Reply:
I think I prefer
```prog_compiled = prog.compile('chip')```
That way the original program is always available (e.g., for comparison)

Reply:
Now compile returns a copy.